design/src/rfcs/rfc0017_logging_sensitive.md

```diff
@@ -0,0 +1,376 @@
+
+Using the smithy model, for each operation, a logging `Layer` should be generated. Through the model, the code generation knows which fields are sensitive and which HTTP bindings exist, therefore the `Logging` layer can be careful crafted to avoid leaking sensitive data.
+
+This `Layer` should be applied to the [OperationHandler](https://github.com/awslabs/smithy-rs/blob/cd0563020abcde866a741fa123e3f2e18e1be1c9/rust-runtime/inlineable/src/server_operation_handler_trait.rs#L17-L21) directly after its construction in the generated `operation_registry.rs`. The `Layer` should preserve the associated types of the `OperationHandler` (`Response = Response<BoxBody>`, `Error = Infallible`) so cause minimal breakage.
+
+This is illustrated before as the introduction of `Logging{Operation}::new`.
+
+```rust
+let empty_operation = LoggingEmptyOperation::new(operation(registry.empty_operation));
+let get_pokemon_species = LoggingPokemonSpecies::new(operation(registry.get_pokemon_species));
+let get_server_statistics = LoggingServerStatistics::new(operation(registry.get_server_statistics));
+let routes = vec![
+    (BoxCloneService::new(empty_operation), empty_operation_request_spec),
+    (BoxCloneService::new(get_pokemon_species), get_pokemon_species_request_spec),
+    (BoxCloneService::new(get_server_statistics), get_server_statistics_request_spec),
+];
+let router = aws_smithy_http_server::routing::Router::new_rest_json_router(routes);
+```
+
+As a request enters this layer it should record the HTTP headers, status code, and URI.
+
+The following model
+
+```smithy
+@readonly
+@http(uri: "/inventory/{name}", method: "GET")
+operation Inventory {
+    input: Product,
+    output: Stocked
+}
+
+@input
+structure Product {
+    @required
+    @sensitive
+    @httpLabel
+    name: String
+}
+
+@output
+structure Stocked {
+    @sensitive
+    @httpResponseCode
+    code: String,
+}
+```
+
+should generate the following
+
+```rust
+// NOTE: This code is intended to show behavior - it does not compile
+
+const SENSITIVE_MARKER: &str = "_redacted_";
+
+pub struct InventoryLogging<S> {
+    inner: S
+}
+
+impl<S> InventoryLogging<S> {
+    pub fn new(inner: S) -> Self {
+        Self {
+            inner
+        }
+    }
+}
+
+impl<B, S> Service<Request<B>> for InventoryLogging<S>
+where
+    S: Service<Request<B>>
+{
+    type Response = Response<BoxBody>;
+    type Error = S::Error;
+    type Future = /* Implementation detail */;
+
+    fn call(&mut self, request: Request<B>) -> Self::Future {
+        // Remove sensitive data from path
+        let mut uri = redact(request.uri());
+
+        let fut = async {
+            let response = self.inner.call(request).await;
+
+            debug!(status_code = %Sensitive(request.status()), "sending response");
+
+            response
+        };
+
+        // Instrument the future with a span
+        let span = span!(%method, %uri, headers = request.headers(), "received request");
```

Comment:
This might be a lot of data to include in a span? We might want a way to reduce the amount placed in here and instead log an `debug!(?headers)` for example.

The benefit of having the `span` is that every log within this will also have access to this data.

Reply:
This is a fair callout. I think I am in favour of have as many infos as possible. What do you think people?